extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -78,10 +78,10 @@ class BindDatesDetails {
     this['dates'] = dates;
 
     /** @const */
-    this['start'] = dates[0];
+    this['start'] = map(dates[0]);
 
     /** @const */
-    this['end'] = dates[dates.length - 1];
+    this['end'] = map(dates[dates.length - 1]);
```

Comment:
For your first question, getBindDates_ always takes at least one date, so the array will always have length 1 as currently written. But it could be good defensive programming practice to add a dev assert.
For your second question, when Bind attempts to `deepMerge` the new state with the old state, the `deepMerge` function marks objects in the new state as `seen` as it iterates through the state tree. If there is an object reference assigned to 2 properties of the new state object, it gets marked as seen and is encountered again, emitting the error. Not sure if this is WAI, but it's easy to work around.
https://github.com/ampproject/amphtml/blob/391a431816894cc726e8c7891081885589f391bf/src/utils/object.js#L105-L107